fish-rust/src/env_dispatch.rs

```diff
@@ -0,0 +1,809 @@
+    crate::fallback::FISH_AMBIGUOUS_WIDTH.store(new_width, Ordering::Relaxed);
+}
+
+fn handle_term_size_change(vars: &dyn Environment) {
+    crate::termsize::handle_columns_lines_var_change(vars);
+}
+
+fn handle_fish_history_change(vars: &dyn Environment) {
+    let fish_history = vars.get(L!("fish_history"));
+    let var = crate::env::env_var_to_ffi(fish_history);
+    crate::ffi::reader_change_history(&crate::ffi::history_session_id(var));
+}
+
+fn handle_fish_cursor_selection_mode_change(vars: &dyn Environment) {
+    use crate::reader::CursorSelectionMode;
+
+    let inclusive = vars
+        .get(L!("fish_cursor_selection_mode"))
+        .as_ref()
+        .map(|v| v.as_string())
+        .map(|v| v == L!("inclusive"))
+        .unwrap_or(false);
+    let mode = if inclusive {
+        CursorSelectionMode::Inclusive
+    } else {
+        CursorSelectionMode::Exclusive
+    };
+
+    let mode = mode as u8;
+    crate::ffi::reader_change_cursor_selection_mode(mode);
+}
+
+fn handle_autosuggestion_change(vars: &dyn Environment) {
+    // TODO: This was a call to reader_set_autosuggestion_enabled(vars) and
+    // reader::check_autosuggestion_enabled() should be private to the `reader` module.
+    crate::ffi::reader_set_autosuggestion_enabled_ffi(crate::reader::check_autosuggestion_enabled(
+        vars,
+    ));
+}
+
+fn handle_function_path_change(_: &dyn Environment) {
+    crate::ffi::function_invalidate_path();
+}
+
+fn handle_complete_path_change(_: &dyn Environment) {
+    crate::ffi::complete_invalidate_path();
+}
+
+fn handle_tz_change(var_name: &wstr, vars: &dyn Environment) {
+    handle_timezone(var_name, vars);
+}
+
+fn handle_locale_change(vars: &dyn Environment) {
+    init_locale(vars);
+    // We need to re-guess emoji width because the locale might have changed to a multibyte one.
+    guess_emoji_width(vars);
+}
+
+fn handle_curses_change(vars: &dyn Environment) {
+    guess_emoji_width(vars);
+    init_curses(vars);
+}
+
+fn handle_fish_use_posix_spawn_change(vars: &dyn Environment) {
+    // Note that if the variable is missing or empty we default to true (if allowed).
+    if !allow_use_posix_spawn() {
+        USE_POSIX_SPAWN.store(false, Ordering::Relaxed);
+    } else if let Some(var) = vars.get(L!("fish_use_posix_spawn")) {
+        let use_posix_spawn =
+            var.is_empty() || crate::wcstringutil::bool_from_string(&var.as_string());
+        USE_POSIX_SPAWN.store(use_posix_spawn, Ordering::Relaxed);
+    } else {
+        USE_POSIX_SPAWN.store(true, Ordering::Relaxed);
+    }
+}
+
+/// Allow the user to override the limits on how much data the `read` command will process. This is
+/// primarily intended for testing, but could also be used directly by users in special situations.
+fn handle_read_limit_change(vars: &dyn Environment) {
+    let read_byte_limit = vars
+        .get_unless_empty(L!("fish_read_limit"))
+        .map(|v| v.as_string())
+        .and_then(|v| match v.to_string().parse() {
```

Comment:
The previous code used `fish_wcstoull` which does something different than `parse`. This should probably use `fish_wcstoi`.

Reply:
I actually intentionally used `usize's` `FromStr` impl to avoid the previous behavior which silently truncated on 32-bit platforms when assigning the `unsigned long long` result to `size_t read_byte_limit`. 

Note that using `fish_wcstoi` would technically be a breaking change as well as read limits over 4GB would become invalid (I don't have a problem with that, but I assumed there was a reason `read_byte_limit` was a `size_t`).

Reply:
It's a `size_t` because it effectively controls the size of a memory buffer. I'm OK expanding it to `u64` or whatever.

But (IMO) the more important difference is that `fish_wcstoull` recognizes a different set of strings than `parse()`. For example, the `fish_` variants skip leading whitespace while `parse` does not. So switching to `parse()` would technically be a breaking change; that is why we should continue to use the `fish_*` set of string parsing functions everywhere they appear in the current C++ code.